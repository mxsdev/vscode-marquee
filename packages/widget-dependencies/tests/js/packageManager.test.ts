import vscode from 'vscode'
import { ChildProcess, exec } from 'child_process'
import projects from './projects'

import {
  tryGetJsProject,
  tryGetPackageJson,
  findWorkspaces,
  packageManagerCmd,
} from '../../src/js/packageManager'

jest.mock('vscode', () => jest.requireActual('../vscodeFsMock.ts'))

jest.mock('child_process', () => ({
  exec: jest.fn()
}))

test('tryGetJsProject', async () => {
  const jsProject = await tryGetJsProject(
    vscode.Uri.file(projects.yarn)
  )

  expect(jsProject).toBeTruthy()
  expect(jsProject!.manager).toBe('yarn')
  expect(Object.keys(jsProject!.dependencies)).toHaveLength(5)

  expect(jsProject).toMatchSnapshot()
})

test('tryGetPackageJson', async () => {
  const packageJson = await tryGetPackageJson(
    vscode.Uri.file(projects.yarn)
  )

  expect(packageJson).toBeTruthy()
  expect(packageJson!.version).toBe('1.0.0')

  expect(packageJson).toMatchSnapshot()
})

test('findWorkspaces', async () => {
  const rootProject = await tryGetJsProject(
    vscode.Uri.file(projects.yarnWorkspaces)
  )

  expect(rootProject).toBeTruthy()

  const workspaces = await findWorkspaces(rootProject!)

  expect(workspaces).toBeTruthy()

  const workspaceNames = Object.keys(workspaces)

  expect(workspaceNames).toHaveLength(4)

  expect(workspaceNames.includes('yarn-project-workspace')).toBeTruthy()
  expect(workspaceNames.includes('yarn-project-workspace-package1')).toBeTruthy()
  expect(workspaceNames.includes('yarn-project-workspace-package2')).toBeTruthy()
  expect(workspaceNames.includes('yarn-project-workspace-package3')).toBeTruthy()
})

describe('packageManagerCmd', () => {
  test('caches result', async () => {
    jest.mocked(exec).mockImplementation((command, opts, callback) => {
      callback?.(null, command, '')
      return { on: jest.fn() } as unknown as ChildProcess
    })

    const pkg = await tryGetJsProject(
      vscode.Uri.file(projects.yarn)
    )

    expect(pkg).toBeTruthy()

    expect(await packageManagerCmd(pkg!, 'test', 'command')).toEqual('yarn test command')
    expect(await packageManagerCmd(pkg!, 'test', 'command')).toEqual('yarn test command')

    expect(jest.mocked(exec)).toBeCalledTimes(1)
  })

  test('caches separate packages separately', async () => {
    jest.mocked(exec).mockImplementation((command, opts, callback) => {
      callback?.(null, command, '')
      return { on: jest.fn() } as unknown as ChildProcess
    })

    {
      const pkg = await tryGetJsProject(
        vscode.Uri.file(projects.yarn)
      )

      expect(pkg).toBeTruthy()

      expect(await packageManagerCmd(pkg!, 'test', 'command')).toEqual('yarn test command')
      expect(await packageManagerCmd(pkg!, 'test', 'command')).toEqual('yarn test command')
    }

    {
      const pkg = await tryGetJsProject(
        vscode.Uri.file(projects.yarnWorkspaces)
      )

      expect(pkg).toBeTruthy()

      expect(await packageManagerCmd(pkg!, 'test', 'command')).toEqual('yarn test command')
      expect(await packageManagerCmd(pkg!, 'test', 'command')).toEqual('yarn test command')
    }

    expect(jest.mocked(exec)).toBeCalledTimes(2)
  })
})
