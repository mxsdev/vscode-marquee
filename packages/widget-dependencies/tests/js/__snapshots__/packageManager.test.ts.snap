// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tryGetJsProject 1`] = `
{
  "dependencies": {
    "@emotion/react": {
      "actualVersion": "11.10.5",
      "dependencyType": "dev",
      "homepage": undefined,
      "name": "@emotion/react",
      "packageVersion": "^11.10.5",
      "repository": "https://github.com/emotion-js/emotion/tree/main/packages/react",
    },
    "@emotion/styled": {
      "actualVersion": "11.10.5",
      "dependencyType": "normal",
      "homepage": undefined,
      "name": "@emotion/styled",
      "packageVersion": "^11.10.5",
      "repository": "https://github.com/emotion-js/emotion/tree/main/packages/styled",
    },
    "@mui/material": {
      "actualVersion": "5.10.13",
      "dependencyType": "dev",
      "homepage": "https://mui.com/material-ui/getting-started/overview/",
      "name": "@mui/material",
      "packageVersion": "^5.10.13",
      "repository": undefined,
    },
    "@types/react": {
      "actualVersion": "18.0.25",
      "dependencyType": "dev",
      "homepage": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react",
      "name": "@types/react",
      "packageVersion": "^18.0.25",
      "repository": undefined,
    },
    "typescript": {
      "actualVersion": "4.9.4",
      "dependencyType": "dev",
      "homepage": "https://www.typescriptlang.org/",
      "name": "typescript",
      "packageVersion": "4.9.4",
      "repository": undefined,
    },
  },
  "homepage": undefined,
  "isWorkspace": false,
  "json": {
    "dependencies": {
      "@emotion/styled": "^11.10.5",
    },
    "devDependencies": {
      "@emotion/react": "^11.10.5",
      "@mui/material": "^5.10.13",
      "@types/react": "^18.0.25",
      "typescript": "4.9.4",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "yarn-project",
    "version": "1.0.0",
  },
  "manager": "yarn",
  "name": "yarn-project",
  "repository": undefined,
  "uri": "packages/widget-dependencies/tests/js/projects/yarn",
  "version": "1.0.0",
  "workspaces": undefined,
}
`;

exports[`tryGetPackageJson 1`] = `
{
  "dependencies": {
    "@emotion/styled": "^11.10.5",
  },
  "devDependencies": {
    "@emotion/react": "^11.10.5",
    "@mui/material": "^5.10.13",
    "@types/react": "^18.0.25",
    "typescript": "4.9.4",
  },
  "license": "MIT",
  "main": "index.js",
  "name": "yarn-project",
  "version": "1.0.0",
}
`;
